aioxmpp/service.py

```diff
@@ -854,29 +855,74 @@ class HandlerSpec(collections.namedtuple(
     If at class definition time any of the dependent classes in `require_deps`
     are not declared using the order attributes (see :class:`Meta`), a
     :class:`TypeError` is raised.
+
+    There is a property to extract the function directly:
+
+    .. autoproperty:: func
     """
 
     def __new__(cls, key, is_unique=True, require_deps=()):
         return super().__new__(cls, is_unique, key, frozenset(require_deps))
 
+    @property
+    def func(self):
+        """
+        The factory of the context manager for this handler.
+
+        .. versionadded:: 0.11
+        """
+        return self.key[0]
+
+
+def _reduce_handler_kwargs(handler_spec, kwargs):
+    if kwargs is None:
+        kwargs = {}
+    else:
+        kwdefaults = handler_spec.func.__kwdefaults__ or {}
+        for key, value in list(kwargs.items()):
+            if key not in kwdefaults:
+                raise ValueError("invalid keyword argument")
+            if value == kwdefaults[key]:
+                del kwargs[key]
+    return kwargs
+
 
-def add_handler_spec(f, handler_spec):
+def add_handler_spec(f, handler_spec, *, kwargs=None):
     """
     Attach a handler specification (see :class:`HandlerSpec`) to a function.
 
     :param f: Function to attach the handler specification to.
     :param handler_spec: Handler specification to attach to the function.
     :type handler_spec: :class:`HandlerSpec`
+    :param kwargs: additional keyword arguments for the apply function
+       carried in the handler spec.
+    :type kwargs: :class:`dict`
+
+    :raises ValueError: if the handler was registered with
+       incompatible keyword arguments before
+    :raises ValueError: if the handler function does not take the
+       specified keyword arguments
 
     This uses a private attribute, whose exact name is an implementation
-    detail. The `handler_spec` is stored in a :class:`set` bound to the
+    detail. The `handler_spec` is stored in a :class:`dict` bound to the
     attribute.
+
+    The additional `kwargs` arguments must be keyword-only arguments of
+    the function. These args are reduced: only non-default arguments are
+    passed on, this allows robust comparison of additional arguements for
+    the comptaibility checking.
```

Comment:
Hmmm… Can we get around this by not making the ``with_send_reply`` argument of the ``_apply_iq_handler`` thing not a keyword argument?

Then it would *always* have to be present in the kwargs dictionary and we don’t need this normalization.

(I prefer this way a lot actually, because the ``_reduce_handler_kwargs`` thing feels dangerously close to implementaiton details to me – although it probably isn’t.)

Reply:
Also, add a versionchanged note.

Reply:
Hmm, that sounds right, as long as the handler specs are created uniformly from the decorators, there is no point in doing the normalization (as they will always include values for all kwargs).

Reply:
Yes.